CompilerSource/compiler/components/write_object_data.cpp

```diff
@@ -639,7 +639,7 @@ static inline void write_object_class_bodies(lang_CPP *lcpp, std::ostream &wto,
 }
 
 static inline string resname(string name) {
-  return name.empty() ? "-1" : name;
+  return (name.empty() || (name == "<undefined>")) ? "-1" : name;
```

Comment:
Please don't do this. Here or in write_room_data. Magic strings are not good, and this one is particularly bad and ugly.

Reply:
It's GMX and maybe YYP that have that special value. How do you propose to change it? (It does fix building some games on emake that also build on master).

Reply:
Please handle this in libGMX; extract the string into a constant (eg `RETARDED_IDEA_NUMBER_95327`) and handle the comparison there. If it's set to `<undefined>`, decline to define it (this is how a typical system denotes that something is not defined).

Reply:
We lack infrastructure necessary to deal with the issue fully the way you suggest. It needs handled for every reference a resource makes to another resource. An object's `parent_name` is an example of such a field that has no metadata marking it as a resource reference, which would be necessary to perform the comparison in libEGM's GMX reader.
https://github.com/enigma-dev/enigma-dev/blob/d31cc9e94eedf918cde745ce41a7bd8879d9c16a/shared/protos/Object.proto#L10
https://github.com/enigma-dev/enigma-dev/blob/d31cc9e94eedf918cde745ce41a7bd8879d9c16a/CommandLine/libEGM/gmx.cpp#L428

I need further direction from you if you want to this changed. What this pull request currently has is actually more correct, technically, at the moment.

Reply:
Seems as though the infrastructure needed to handle that is relatively minor (if a little obnoxious), and it's probably better that we have it, anyway, so we can do other validations on it.

Reply:
Well I can only think of two ways to solve it, and would like to know which you prefer.
* Add a `[(ref)=true]` field option that "flags" the field as being a resource reference. This would allow me to know which fields to check the special value `"<undefined>"` for.
* Make an explicit whitelist of fields in the GMX reader that it knows to check for the special value.

Reply:
You're close to what I was thinking; I prefer something more like `[(resource_ref)="sprite"]`. This lets us validate the refs as we load (or shortly thereafter).

But for now, a post-processing step is fine... and I suppose, as a temporary hack, you can do it in the GameData ctor, since that's fully-expanded. You can also go ahead with that feature now, but please do so in its own PR.

Reply:
Alright, so thanks to #1547, we now have that infrastructure I was talking about.

Reply:
Ok: ebba3b0ebcc3fe20bfc19c00fcc2ed09ac45f858

It's been dealt with in the GMX reader now. Testing GMX projects that are affected, like the AI Tutorial, shows no sign of regression. I guess this is resolved.